### PR TITLE
Move LogStorageAppenderTest back to logstreams

### LINKED ISSUE #9244: Rewrite LogStorageAppenderTest
**Description**

Currently, the `LogStorageAppenderTest` is tightly coupled with implementation details. During moving the AtomixLogStorage https://github.com/camunda/zeebe/pull/9235 we had to move the test as well and open the Writer implementation, which is sub-optimal. 

We have to reevaluate again what we actually want to test here. The inconsistency check? Which is part of the ZeebeValidator and LeaderAppender? Then I guess we can test this differently, no need to use the LogStorageAppender here. 

Potentially we can split up the test class and move parts back to the Logstreams tests and leave some in the broker to test the inconsistency check. 

Related comment https://github.com/camunda/zeebe/pull/9235#discussion_r859694253

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

This PR moves the `LogStorageAppenderTest` back into the `logstreams` module, as it is implementation detail of the `logstreams`. It also updates it to use JUnit 5.

The inconsistency test is rewritten, as the check was also modified. The `ZeebeEntryValidator` now only checks if there are no gaps between `ApplicationEntry`, using the `lowestPosition` and `highestPosition` properties. This is because its previous implementation, using the `LogEntryDescriptor`, was using `logstreams` implementation specific information.

Instead, checking if there are no gaps in a batch is performed in the `LogStorageAppender` directly. There, if there are gaps between entries, the actor is failed. This is propagated upwards until the partition fails and steps down.

I also did a bit of clean up and maintenance, but nothing that's expected to have a major impact. I will try a spike and do a follow up here to split the `LogStorageAppender` into smaller, single responsibility classes, in order to make maintenance and testing easier. For now I would close the issue as is though.

## Related issues

closes #9244 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by npepinpe** [comment]
It seems like a lot of changes, but most of them come from the following two structural changes:

- Transformed Raft and Journal data classes into records, which greatly cut down on the boilerplate.
- Nested `NoopEntryValidator` and `ValidationResult` into the `EntryValidator` interface. My reasoning here is both of these are only meant to be used by implementations of the validator and only within that context. I wouldn't want `ValidationResult`, for example, to start being used for other types of validation as it's really not meant to be (at least not without some refactoring I guess).

So you can focus on the following four commits (in order):

- https://github.com/camunda/zeebe/commit/1b4bc5a9dd1aeb9b4beba61242cb0a518e9c6684
- https://github.com/camunda/zeebe/commit/5460d1c2307a49a402a403e680f7536afb829506
- https://github.com/camunda/zeebe/commit/b41edddb5cf40e588e7ce7b3f4fb31c3853b726e
- https://github.com/camunda/zeebe/commit/14db7410f834facc3b82d42bf335518b582902e6

File: logstreams/src/test/java/io/camunda/zeebe/logstreams/impl/log/LogStorageAppenderTest.java [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+import java.util.concurrent.atomic.AtomicReference;
+import org.agrona.CloseHelper;
+import org.agrona.MutableDirectBuffer;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.parallel.Execution;
+import org.junit.jupiter.api.parallel.ExecutionMode;
+
+@Execution(ExecutionMode.CONCURRENT)
```
Comment (deepthidevaki) [inline logstreams/src/test/java/io/camunda/zeebe/logstreams/impl/log/LogStorageAppenderTest.java:49]: :question: Is this something that we want to do for all tests? Is there any impact on setting this?

**Review by deepthidevaki — APPROVED** [review]
Thanks, looks good :+1: 

Just one minor comment. No need for another review.

File: logstreams/src/test/java/io/camunda/zeebe/logstreams/impl/log/LogStorageAppenderTest.java (hunk 1 above)
Comment (npepinpe) [reply logstreams/src/test/java/io/camunda/zeebe/logstreams/impl/log/LogStorageAppenderTest.java:49]: There is an impact: each tests run in a different thread. This is fine for self-contained, lightweight tests. This is an issue for more complex ones, mostly due to increased load. You also have to ensure your tests share no state at all, which hopefully is already the case :smile:

File: logstreams/src/test/java/io/camunda/zeebe/logstreams/impl/log/LogStorageAppenderTest.java [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+      assertThat(reader.hasNext()).isTrue();
+      assertThat(Value.of(reader.next())).isEqualTo(value);
+    }
+  }
+
+  @Test
+  void shouldFailActorWhenDetectingGapsInPositions() throws InterruptedException {
+    // given
+    final var value = new Value(1);
+    final var committed = new CountDownLatch(1);
```
[the remark this answers is not in the document]
Comment (npepinpe) [reply logstreams/src/test/java/io/camunda/zeebe/logstreams/impl/log/LogStorageAppenderTest.java:156]: Nice catch :+1:

**Comment by npepinpe** [comment]
bors merge

**Pull request merged**